File: Sample/Engine/UI/Base/FunctionalBase.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
import logging
from typing import Callable, Dict, Any, Optional
from ... import FloatRect
from ...Utils import Math
from .FocusableMixin import Direction, FocusableMixin


class FunctionalBase(FocusableMixin):
    """Mixin providing interactive event callbacks for UI controls.

    Handles hover detection and dispatches confirm, cancel, click, hover,
    mouse, keyboard, and tick events to registered callbacks.
    """

    _keyboardFocusResolver: Optional[Callable[["FunctionalBase"], bool]] = None
    _directionalFocusRequester: Optional[Callable[["FunctionalBase", Direction], bool]] = None
    _keyboardFocusSetter: Optional[Callable[["FunctionalBase"], bool]] = None
    _keyboardCursorResolver: Optional[Callable[["FunctionalBase"], bool]] = None

    def __init__(self) -> None:
        r"""\brief Construct a FunctionalBase mixin.

        Initialises the hovered state to False.
        """
        FocusableMixin.__init__(self)
        self._isHovered: bool = False
        self._active: bool = True

    @classmethod
    def setKeyboardFocusResolver(
        cls,
        resolver: Optional[Callable[["FunctionalBase"], bool]],
    ) -> None:
        r"""\brief Set the global keyboard focus resolver.

        - \param resolver Callback returning whether a control receives keyboard events.
        """
        cls._keyboardFocusResolver = resolver

    @classmethod
    def setDirectionalFocusRequester(
        cls,
        requester: Optional[Callable[["FunctionalBase", Direction], bool]],
    ) -> None:
        r"""\brief Set the global directional focus move requester.

        - \param requester Callback used by controls when focus should leave them.
        """
        cls._directionalFocusRequester = requester

    @classmethod
    def setKeyboardFocusSetter(
        cls,
        setter: Optional[Callable[["FunctionalBase"], bool]],
    ) -> None:
        r"""\brief Set the global keyboard focus setter.

        - \param setter Callback used by controls to request keyboard focus.
        """
        cls._keyboardFocusSetter = setter

    @classmethod
    def setKeyboardCursorResolver(
        cls,
        resolver: Optional[Callable[["FunctionalBase"], bool]],
    ) -> None:
        r"""\brief Set the global keyboard cursor owner resolver.

        - \param resolver Callback returning whether a control owns the selection cursor.
        """
        cls._keyboardCursorResolver = resolver

    def canReceiveFocus(self) -> bool:
        r"""\brief Return whether this control can currently receive focus.

        - \return True if the control is active, visible when applicable, and focusable.
        """
        from .ControlBase import ControlBase

        if not self.getCanReceiveFocus() or not self.getActive():
            return False
        if isinstance(self, ControlBase) and not self.getVisible():
            return False
        return True

    def shouldDispatchKeyboardInput(self) -> bool:
        r"""\brief Return whether keyboard callbacks should be dispatched.

        - \return True if no focus resolver blocks this control.
        """
        resolver = type(self)._keyboardFocusResolver
        if resolver is None:
            return True
        return resolver(self)

    def requestDirectionalFocusMove(self, direction: Direction) -> bool:
        r"""\brief Request a focus move in a direction.

        - \param direction Navigation direction.
        - \return True if focus moved.
        """
        requester = type(self)._directionalFocusRequester
        if requester is None:
            return False
        return requester(self, direction)

    def requestKeyboardFocus(self) -> bool:
        r"""\brief Request keyboard focus for this control.

        - \return True if focus moved to this control.
        """
        setter = type(self)._keyboardFocusSetter
        if setter is None:
            return False
        return setter(self)

    def ownsKeyboardCursorFocus(self) -> bool:
        r"""\brief Return whether this control owns the selection cursor.

        - \return True if the current focus manager assigns cursor ownership to this control.
        """
        resolver = type(self)._keyboardCursorResolver
        if resolver is None:
            return self.getFocused()
        return resolver(self)

    def isHovered(self) -> bool:
        r"""\brief Check whether the mouse is hovering over this control.

        - \return  True if hovered, False otherwise
        """
        return self._isHovered

    def getActive(self) -> bool:
        r"""\brief Check whether this control is active.

        - \return  True if active, False otherwise
        """
        return self._active

    def setActive(self, active: bool) -> None:
        r"""\brief Set the activity state of this control.

        - \param active  True to activate, False to deactivate
        """
        self._active = active

    def onConfirm(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when the control is confirmed (e.g. mouse click).

        - \param kwargs  Event arguments (may include position)
        """
        pass

    def onCancel(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when the control is cancelled.

        - \param kwargs  Event arguments
        """
        pass

    def onClick(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when the control is clicked.

        - \param kwargs  Event arguments (may include position)
        """
        pass

    def onMouseButtonDown(self, kwargs: Dict[str, Any]) -> bool:
        r"""\brief Called when any mouse button is pressed, regardless of hover state.

        - \param kwargs  Event arguments (includes position and button)

        - \return True if the event was handled and should be consumed.
        """
        return False

    def onHover(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when the mouse starts hovering over the control.

        - \param kwargs  Event arguments (may include position)
        """
        pass

    def onUnHover(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when the mouse stops hovering over the control.

        - \param kwargs  Event arguments (may include position)
        """
        pass

    def onMouseMoved(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when the mouse moves while hovering.

        - \param kwargs  Event arguments (may include position)
        """
        pass

    def onMouseWheelScrolled(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when the mouse wheel is scrolled while hovering.

        - \param kwargs  Event arguments (may include position and delta)
        """
        pass

    def onKeyDown(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when a key or button is pressed while hovering.

        - \param kwargs  Event arguments (empty dict)
        """
        pass

    def onKeyUp(self, kwargs: Dict[str, Any]) -> None:
        r"""\brief Called when a key or button is released while hovering.

        - \param kwargs  Event arguments (empty dict)
        """
        pass

    def onTick(self, deltaTime: float) -> None:
        r"""\brief Called every frame with the frame time.

        - \param deltaTime  Time elapsed since last frame, in seconds
        """
        pass

    def onLateTick(self, deltaTime: float) -> None:
        r"""\brief Called after onTick every frame.

        - \param deltaTime  Time elapsed since last frame, in seconds
        """
        pass

    def onFixedTick(self, fixedDelta: float) -> None:
        r"""\brief Called at a fixed timestep.

        - \param fixedDelta  Fixed timestep duration, in seconds
        """
        pass

    def addConfirmCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the confirm action.

        - \param callback_  Callback to invoke on confirm
        """
        self.onConfirm = callback_.__get__(self, type(self))

    def addCancelCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the cancel action.

        - \param callback_  Callback to invoke on cancel
        """
        self.onCancel = callback_.__get__(self, type(self))

    def addClickCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the click action.

        - \param callback_  Callback to invoke on click
        """
        self.onClick = callback_.__get__(self, type(self))

    def addMouseButtonDownCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the mouse-button-down action.

        - \param callback_  Callback to invoke on mouse button down
        """
        self.onMouseButtonDown = callback_.__get__(self, type(self))

    def addHoverCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the hover action.

        - \param callback_  Callback to invoke on hover
        """
        self.onHover = callback_.__get__(self, type(self))

    def addUnHoverCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the un-hover action.

        - \param callback_  Callback to invoke on un-hover
        """
        self.onUnHover = callback_.__get__(self, type(self))

    def addMouseMovedCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the mouse-moved action.

        - \param callback_  Callback to invoke on mouse move
        """
        self.onMouseMoved = callback_.__get__(self, type(self))

    def addMouseWheelScrolledCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the mouse-wheel-scrolled action.

        - \param callback_  Callback to invoke on mouse wheel scroll
        """
        self.onMouseWheelScrolled = callback_.__get__(self, type(self))

    def addKeyDownCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the key-down action.

        - \param callback_  Callback to invoke on key down
        """
        self.onKeyDown = callback_.__get__(self, type(self))

    def addKeyUpCallback(self, callback_: Callable) -> None:
        r"""\brief Register a callback for the key-up action.

        - \param callback_  Callback to invoke on key up
        """
        self.onKeyUp = callback_.__get__(self, type(self))

    def update(self, deltaTime: float) -> None:
        r"""\brief Process input and fire tick callbacks.

        - \param deltaTime  Time elapsed since last frame, in seconds
        """
        from Engine import Input
        from Engine.UI.Base import ControlBase

        self.onTick(deltaTime)
        if isinstance(self, ControlBase) and not self.getVisible():
            if self._isHovered:
                self._isHovered = False
                self.onUnHover({"position": Math.ToVector2f(Input.getMousePosition())})
            return
        localMousePos = Math.ToVector2f(Input.getMousePosition())
        active = self.getActive()
        if active and Input.isMouseButtonPressed():
            for btn in [Input.Mouse.Button.Left, Input.Mouse.Button.Right, Input.Mouse.Button.Middle]:
                if not Input.getMouseButtonPressed(btn, handled=False):
                    continue
                if self.onMouseButtonDown({"position": localMousePos, "button": btn}):
                    Input.getMouseButtonPressed(btn, handled=True)
                    Input.isMouseButtonTriggered(btn, handled=True)
        hovered = False
        if isinstance(self, ControlBase):
            bounds: FloatRect = self.getAbsoluteBounds()
            hovered = bounds.contains(localMousePos)
        if not Input.isMouseInputMode():
            hovered = False
        if hovered:
            if not self._isHovered:
                self._isHovered = True
                self.onHover({"position": localMousePos})
            if Input.isMouseMoved():
                self.onMouseMoved({"position": localMousePos})
            if active and Input.isMouseButtonPressed():
                self.onClick({"position": localMousePos})
            if active and Input.isMouseWheelScrolled():
                self.onMouseWheelScrolled({"position": localMousePos, "delta": Input.getMouseScrolledWheelDelta()})
        if not hovered:
            if self._isHovered:
                self._isHovered = False
                self.onUnHover({"position": localMousePos})
        if active and isinstance(self, ControlBase):
            bounds = self.getAbsoluteBounds()
            if Input.isTouchBegan():
                beganPos = Input.getTouchBeganPosition()
                if beganPos is not None:
                    touchLocal = Math.ToVector2f(beganPos)
                    if bounds.contains(touchLocal):
                        self.onClick({"position": touchLocal})
            if Input.isTouchMoved():
                touchPos = Input.getTouchPosition()
                if touchPos is not None:
                    touchLocal = Math.ToVector2f(touchPos)
                    if bounds.contains(touchLocal):
                        self.onMouseMoved({"position": touchLocal})
        if self.shouldDispatchKeyboardInput():
            if Input.isKeyPressed() or Input.isJoystickButtonPressed() or Input.isJoystickAxisMoved():
                self.onKeyDown({})
            if Input.isKeyReleased() or Input.isJoystickButtonReleased():
                self.onKeyUp({})

    def lateUpdate(self, deltaTime: float) -> None:
        r"""\brief Run late-update tick callback.

        - \param deltaTime  Time elapsed since last frame, in seconds
        """
        self.onLateTick(deltaTime)

    def fixedUpdate(self, fixedDelta: float) -> None:
        r"""\brief Run fixed-timestep tick callback.

        - \param fixedDelta  Fixed timestep duration, in seconds
        """
        self.onFixedTick(fixedDelta)

    def __del__(self) -> None:
        r"""\brief Destructor for FunctionalBase.

        Logs at debug level when the object is collected.
        """
        try:
            super().__del__()
        except AttributeError:
            pass
        logging.debug(f"FunctionalBase {self} deleted")

File: Sample/Core/Engine/src/UI/FunctionalBase.cpp
#include <UI/FunctionalBase.hpp>

#include "FunctionalBase/InteractionRuntime.hpp"

#include <Runtime/EngineState.hpp>
#include <UI/ControlBase.hpp>

#include <array>

FunctionalInputProvider* FunctionalBase::inputProvider_ = nullptr;
FunctionalBase::FocusResolver FunctionalBase::keyboardFocusResolver_;
FunctionalBase::DirectionalFocusRequester
    FunctionalBase::directionalFocusRequester_;
FunctionalBase::FocusSetter FunctionalBase::keyboardFocusSetter_;
FunctionalBase::FocusResolver FunctionalBase::keyboardCursorResolver_;

FunctionalBase::~FunctionalBase() {
    if (pointerSource_ == PointerSource::Touch && inputProvider_ != nullptr) {
        inputProvider_->cancelTouchGesture();
    }
}

void FunctionalBase::setInputProvider(FunctionalInputProvider* provider) {
    inputProvider_ = provider;
}

void FunctionalBase::setKeyboardFocusResolver(FocusResolver resolver) {
    keyboardFocusResolver_ = std::move(resolver);
}

void FunctionalBase::setDirectionalFocusRequester(
    DirectionalFocusRequester requester) {
    directionalFocusRequester_ = std::move(requester);
}

void FunctionalBase::setKeyboardFocusSetter(FocusSetter setter) {
    keyboardFocusSetter_ = std::move(setter);
}

void FunctionalBase::setKeyboardCursorResolver(FocusResolver resolver) {
    keyboardCursorResolver_ = std::move(resolver);
}

void FunctionalBase::resetRuntimeCallbacks() noexcept {
    inputProvider_ = nullptr;
    keyboardFocusResolver_ = {};
    directionalFocusRequester_ = {};
    keyboardFocusSetter_ = {};
    keyboardCursorResolver_ = {};
}

bool FunctionalBase::canReceiveFocus() const {
    return getCanReceiveFocus() && isInteractionEnabled();
}

bool FunctionalBase::shouldDispatchKeyboardInput() const {
    return isInteractionEnabled() &&
           (!keyboardFocusResolver_ || keyboardFocusResolver_(*this));
}

bool FunctionalBase::requestDirectionalFocusMove(const std::string& direction) {
    return isInteractionEnabled() && directionalFocusRequester_ &&
           directionalFocusRequester_(*this, direction);
}

bool FunctionalBase::requestKeyboardFocus() {
    return isInteractionEnabled() && keyboardFocusSetter_ &&
           keyboardFocusSetter_(*this);
}

bool FunctionalBase::ownsKeyboardCursorFocus() const {
    return isInteractionEnabled() &&
           (keyboardCursorResolver_ ? keyboardCursorResolver_(*this)
                                    : getFocused());
}

bool FunctionalBase::isHovered() const {
    return hovered_;
}

bool FunctionalBase::isPressed() const {
    return pressed_;
}

bool FunctionalBase::getActive() const {
    return active_;
}

void FunctionalBase::setActive(bool active) {
    if (active_ == active) {
        return;
    }
    active_ = active;
    if (!active_) {
        const bool hadPointerInteraction =
            hovered_ || pressed_ || pointerSource_ != PointerSource::None;
        ControlBase* control = dynamic_cast<ControlBase*>(this);
        if (control != nullptr) {
            ControlBase::resetFunctionalInteractions(*control);
        } else {
            resetPointerInteraction();
        }
        if (!hadPointerInteraction) {
            onInteractionStateChanged();
        }
        return;
    }
    onInteractionStateChanged();
}

void FunctionalBase::setTouchHitBounds(
    const std::optional<sf::FloatRect>& bounds) {
    ludork::engine::functional_base_impl::validateTouchHitBounds(bounds);
    touchHitBounds_ = bounds;
}

std::optional<sf::FloatRect> FunctionalBase::getAbsoluteTouchHitBounds() const {
    const ControlBase* control = dynamic_cast<const ControlBase*>(this);
    if (control == nullptr) {
        return std::nullopt;
    }
    if (!touchHitBounds_.has_value()) {
        return control->getAbsoluteBounds();
    }
    const sf::FloatRect scaledBounds(touchHitBounds_->position * Scale,
                                     touchHitBounds_->size * Scale);
    return control->screenRenderTransform().transformRect(scaledBounds);
}

void FunctionalBase::addConfirmCallback(EventCallback callback) {
    confirmCallback_ = std::move(callback);
}

void FunctionalBase::addCancelCallback(EventCallback callback) {
    cancelCallback_ = std::move(callback);
}

void FunctionalBase::addClickCallback(EventCallback callback) {
    clickCallback_ = std::move(callback);
}

void FunctionalBase::addMouseButtonDownCallback(HandledEventCallback callback) {
    mouseButtonDownCallback_ = std::move(callback);
}

void FunctionalBase::addHoverCallback(EventCallback callback) {
    hoverCallback_ = std::move(callback);
}

void FunctionalBase::addUnHoverCallback(EventCallback callback) {
    unHoverCallback_ = std::move(callback);
}

void FunctionalBase::addMouseMovedCallback(EventCallback callback) {
    mouseMovedCallback_ = std::move(callback);
}

void FunctionalBase::addMouseWheelScrolledCallback(EventCallback callback) {
    mouseWheelScrolledCallback_ = std::move(callback);
}

void FunctionalBase::addKeyDownCallback(EventCallback callback) {
    keyDownCallback_ = std::move(callback);
}

void FunctionalBase::addKeyUpCallback(EventCallback callback) {
    keyUpCallback_ = std::move(callback);
}

void FunctionalBase::clearEventCallbacks() noexcept {
    confirmCallback_ = {};
    cancelCallback_ = {};
    clickCallback_ = {};
    mouseButtonDownCallback_ = {};
    hoverCallback_ = {};
    unHoverCallback_ = {};
    mouseMovedCallback_ = {};
    mouseWheelScrolledCallback_ = {};
    keyDownCallback_ = {};
    keyUpCallback_ = {};
}

void FunctionalBase::update(float deltaTime) {
    onTick(deltaTime);
    if (inputProvider_ == nullptr) {
        resetPointerInteraction();
        return;
    }
    ControlBase* control = dynamic_cast<ControlBase*>(this);
    const sf::Vector2f mousePosition(inputProvider_->getMousePosition());
    if (!isInteractionEnabled()) {
        resetPointerInteraction();
        return;
    }

    if (control != nullptr && pointerSource_ == PointerSource::None &&
        acceptsTouchCapture()) {
        const std::optional<sf::FloatRect> touchBounds =
            getAbsoluteTouchHitBounds();
        if (touchBounds.has_value() && inputProvider_->isTouchBegan(false)) {
            const std::optional<sf::Vector2i> beganPosition =
                inputProvider_->getTouchBeganPosition();
            if (beganPosition.has_value() &&
                touchBounds->contains(sf::Vector2f(*beganPosition))) {
                beginTouchPress();
                inputProvider_->isTouchBegan(true);
                onTouchCaptureBegan(sf::Vector2f(*beganPosition));
            }
        }
    }
    if (!isInteractionEnabled()) {
        resetPointerInteraction();
        return;
    }

    if (pointerSource_ != PointerSource::Touch) {
        static constexpr std::array buttons = {
            sf::Mouse::Button::Left,
            sf::Mouse::Button::Right,
            sf::Mouse::Button::Middle,
        };
        std::array<bool, buttons.size()> mousePressed = {};
        bool mousePressReceived = false;
        if (inputProvider_->isMouseButtonPressed()) {
            for (std::size_t index = 0; index < buttons.size(); ++index) {
                if (!isInteractionEnabled()) {
                    break;
                }
                const sf::Mouse::Button button = buttons[index];
                mousePressed[index] =
                    inputProvider_->getMouseButtonPressed(button, false);
                mousePressReceived = mousePressReceived || mousePressed[index];
                if (mousePressed[index] &&
                    onMouseButtonDown(
                        mouseButtonArguments(mousePosition, button))) {
                    inputProvider_->getMouseButtonPressed(button, true);
                    inputProvider_->isMouseButtonTriggered(button, true);
                }
            }
        }
        if (!isInteractionEnabled()) {
            resetPointerInteraction();
            return;
        }

        bool hovered = control != nullptr &&
                       control->getAbsoluteBounds().contains(mousePosition);
        if (!inputProvider_->isMouseInputMode()) {
            hovered = false;
        }
        setHovered(hovered, mousePosition);
        if (!isInteractionEnabled()) {
            resetPointerInteraction();
            return;
        }
        if (hovered) {
            if (pointerSource_ == PointerSource::None) {
                for (std::size_t index = 0; index < buttons.size(); ++index) {
                    if (mousePressed[index]) {
                        beginMousePress(buttons[index]);
                        break;
                    }
                }
            }
            if (!isInteractionEnabled()) {
                resetPointerInteraction();
                return;
            }
            if (inputProvider_->isMouseMoved()) {
                onMouseMoved(pointerArguments(mousePosition));
            }
            if (!isInteractionEnabled()) {
                resetPointerInteraction();
                return;
            }
            if (mousePressReceived) {
                onClick(pointerArguments(mousePosition));
            }
            if (!isInteractionEnabled()) {
                resetPointerInteraction();
                return;
            }
            if (inputProvider_->isMouseWheelScrolled()) {
                onMouseWheelScrolled(mouseWheelArguments(
                    mousePosition,
                    inputProvider_->getMouseScrolledWheelDelta()));
            }
            if (!isInteractionEnabled()) {
                resetPointerInteraction();
                return;
            }
        }

        if (pointerSource_ == PointerSource::Mouse) {
            const sf::Mouse::Button button = *pressedMouseButton_;
            const bool released =
                inputProvider_->isMouseButtonReleased() &&
                inputProvider_->getMouseButtonReleased(button, false);
            if (!isInteractionEnabled() || !hovered || released ||
                !inputProvider_->isMouseButtonDown(button)) {
                endPointerPress();
            }
        }
    } else {
        setHovered(false, mousePosition);
    }
    if (!isInteractionEnabled()) {
        resetPointerInteraction();
        return;
    }

    if (control != nullptr) {
        if (pointerSource_ == PointerSource::Touch) {
            const bool ended = inputProvider_->isTouchEnded();
            const std::optional<sf::Vector2i> position =
                ended ? inputProvider_->getTouchEndedPosition()
                      : inputProvider_->getTouchPosition();
            if (inputProvider_->isTouchMoved() && position.has_value()) {
                onMouseMoved(pointerArguments(sf::Vector2f(*position)));
            }
            if (!isInteractionEnabled()) {
                resetPointerInteraction();
                return;
            }
            if (ended) {
                if (inputProvider_->isTouchTap(false)) {
                    const std::optional<sf::FloatRect> releaseBounds =
                        getAbsoluteTouchHitBounds();
                    if (position.has_value() && releaseBounds.has_value() &&
                        releaseBounds->contains(sf::Vector2f(*position))) {
                        const RuntimeValue::Map arguments =
                            pointerArguments(sf::Vector2f(*position));
                        if (confirmCallback_) {
                            onConfirm(arguments);
                        } else {
                            onClick(arguments);
                        }
                    }
                    inputProvider_->isTouchTap(true);
                }
                endPointerPress();
            } else if (!isInteractionEnabled() ||
                       !inputProvider_->isTouchActive() ||
                       !position.has_value()) {
                resetPointerInteraction();
            }
        }
    } else if (pointerSource_ == PointerSource::Touch) {
        resetPointerInteraction();
    }

    if (shouldDispatchKeyboardInput()) {
        if (inputProvider_->isKeyPressed() ||
            inputProvider_->isJoystickButtonPressed() ||
            inputProvider_->isJoystickAxisMoved()) {
            onKeyDown({});
        }
        if (shouldDispatchKeyboardInput() &&
            (inputProvider_->isKeyReleased() ||
             inputProvider_->isJoystickButtonReleased())) {
            onKeyUp({});
        }
    }
}

void FunctionalBase::lateUpdate(float deltaTime) {
    onLateTick(deltaTime);
}

void FunctionalBase::fixedUpdate(float fixedDelta) {
    onFixedTick(fixedDelta);
}

void FunctionalBase::onConfirm(const RuntimeValue::Map& arguments) {
    if (confirmCallback_) {
        confirmCallback_(*this, arguments);
    }
}

void FunctionalBase::onCancel(const RuntimeValue::Map& arguments) {
    if (cancelCallback_) {
        cancelCallback_(*this, arguments);
    }
}

void FunctionalBase::onClick(const RuntimeValue::Map& arguments) {
    if (clickCallback_) {
        clickCallback_(*this, arguments);
    }
}

bool FunctionalBase::onMouseButtonDown(const RuntimeValue::Map& arguments) {
    return mouseButtonDownCallback_ &&
           mouseButtonDownCallback_(*this, arguments);
}

void FunctionalBase::onHover(const RuntimeValue::Map& arguments) {
    if (hoverCallback_) {
        hoverCallback_(*this, arguments);
    }
}

void FunctionalBase::onUnHover(const RuntimeValue::Map& arguments) {
    if (unHoverCallback_) {
        unHoverCallback_(*this, arguments);
    }
}

void FunctionalBase::onMouseMoved(const RuntimeValue::Map& arguments) {
    if (mouseMovedCallback_) {
        mouseMovedCallback_(*this, arguments);
    }
}

void FunctionalBase::onMouseWheelScrolled(const RuntimeValue::Map& arguments) {
    if (mouseWheelScrolledCallback_) {
        mouseWheelScrolledCallback_(*this, arguments);
    }
}

void FunctionalBase::onKeyDown(const RuntimeValue::Map& arguments) {
    if (keyDownCallback_) {
        keyDownCallback_(*this, arguments);
    }
}

void FunctionalBase::onKeyUp(const RuntimeValue::Map& arguments) {
    if (keyUpCallback_) {
        keyUpCallback_(*this, arguments);
    }
}

void FunctionalBase::onTick(float) {}

void FunctionalBase::onLateTick(float) {}

void FunctionalBase::onFixedTick(float) {}

FunctionalInputProvider* FunctionalBase::inputProvider() {
    return inputProvider_;
}

bool FunctionalBase::isInteractionEnabled() const {
    if (!active_) {
        return false;
    }
    const ControlBase* control = dynamic_cast<const ControlBase*>(this);
    if (control == nullptr) {
        return true;
    }
    if (!control->getVisible()) {
        return false;
    }
    std::shared_ptr<ControlBase> parent = control->getParent();
    while (parent != nullptr) {
        if (!parent->getVisible()) {
            return false;
        }
        const FunctionalBase* functional =
            dynamic_cast<const FunctionalBase*>(parent.get());
        if (functional != nullptr && !functional->getActive()) {
            return false;
        }
        parent = parent->getParent();
    }
    return true;
}

bool FunctionalBase::acceptsTouchCapture() const {
    return getCanReceiveFocus() || static_cast<bool>(confirmCallback_) ||
           static_cast<bool>(clickCallback_);
}

void FunctionalBase::onTouchCaptureBegan(const sf::Vector2f&) {}

bool FunctionalBase::hasTouchCapture() const {
    return pointerSource_ == PointerSource::Touch;
}

void FunctionalBase::onPointerInteractionReset() {}

void FunctionalBase::resetPointerInteraction() {
    if (ludork::engine::functional_base_impl::pointerStateIsEmpty(
            hovered_, pressed_, pointerSource_ != PointerSource::None)) {
        onPointerInteractionReset();
        return;
    }
    if (pointerSource_ == PointerSource::Touch && inputProvider_ != nullptr) {
        inputProvider_->cancelTouchGesture();
    }
    hovered_ = false;
    pressed_ = false;
    pointerSource_ = PointerSource::None;
    pressedMouseButton_.reset();
    onPointerInteractionReset();
    onInteractionStateChanged();
}

void FunctionalBase::onInteractionStateChanged() {}

RuntimeValue::Map FunctionalBase::pointerArguments(
    const sf::Vector2f& position) {
    return ludork::engine::functional_base_impl::pointerArguments(position);
}

RuntimeValue::Map FunctionalBase::mouseButtonArguments(
    const sf::Vector2f& position, sf::Mouse::Button button) {
    return ludork::engine::functional_base_impl::mouseButtonArguments(position,
                                                                      button);
}

RuntimeValue::Map FunctionalBase::mouseWheelArguments(
    const sf::Vector2f& position, float delta) {
    return ludork::engine::functional_base_impl::mouseWheelArguments(position,
                                                                     delta);
}

void FunctionalBase::setHovered(bool hovered, const sf::Vector2f& position) {
    if (hovered_ == hovered) {
        return;
    }
    hovered_ = hovered;
    onInteractionStateChanged();
    if (hovered_) {
        onHover(pointerArguments(position));
    } else {
        onUnHover(pointerArguments(position));
    }
}

void FunctionalBase::beginMousePress(sf::Mouse::Button button) {
    pressed_ = true;
    pointerSource_ = PointerSource::Mouse;
    pressedMouseButton_ = button;
    onInteractionStateChanged();
}

void FunctionalBase::beginTouchPress() {
    pressed_ = true;
    pointerSource_ = PointerSource::Touch;
    pressedMouseButton_.reset();
    onInteractionStateChanged();
}

void FunctionalBase::endPointerPress() {
    if (!pressed_) {
        return;
    }
    pressed_ = false;
    pointerSource_ = PointerSource::None;
    pressedMouseButton_.reset();
    onInteractionStateChanged();
}
